src/WebJobs.Script.WebHost/Security/KeyManagement/DefaultSecretsRepositoryFactory.cs

```diff
@@ -5,26 +5,68 @@
         {
-            string secretStorageType = settingsManager.GetSetting(EnvironmentSettingNames.AzureWebJobsSecretStorageType);
-            string storageString = AmbientConnectionStringProvider.Instance.GetConnectionString(ConnectionStringNames.Storage);
-            if (secretStorageType != null && secretStorageType.Equals("Blob", StringComparison.OrdinalIgnoreCase) && storageString != null)
+            // 3 attempts to copy keys from file system to storage
+            for (int i = 0; i < 3; i++)
             {
+                string secretStorageType = settingsManager.GetSetting(EnvironmentSettingNames.AzureWebJobsSecretStorageType);
+                string storageString = AmbientConnectionStringProvider.Instance.GetConnectionString(ConnectionStringNames.Storage);
                 string siteSlotName = settingsManager.AzureWebsiteUniqueSlotName ?? config.HostConfig.HostId;
-                return new BlobStorageSecretsRepository(Path.Combine(webHostSettings.SecretsPath, "Sentinels"), storageString, siteSlotName);
+                if (secretStorageType != null && secretStorageType.Equals("Blob", StringComparison.OrdinalIgnoreCase) && storageString != null)
+                {
+                    return new BlobStorageSecretsRepository(Path.Combine(webHostSettings.SecretsPath, "Sentinels"), storageString, siteSlotName);
+                }
+                else
+                {
+                    if (string.IsNullOrEmpty(storageString))
+                    {
+                        return new FileSystemSecretsRepository(webHostSettings.SecretsPath);
+                    }
+                    try
+                    {
+                        // Change to blob storage repository
+                        CopyKeysFromFileSystemToBlobStorage(webHostSettings.SecretsPath, storageString, siteSlotName).Wait();
+                        settingsManager.SetSetting(EnvironmentSettingNames.AzureWebJobsSecretStorageType, "Blob");
+                    }
+                    catch
+                    {
+                        // best effort
+                    }
+                }
             }
-            else
+
+            return new FileSystemSecretsRepository(webHostSettings.SecretsPath);
+        }
+
+        private async Task CopyKeysFromFileSystemToBlobStorage(string secretsPath, string storageString, string siteSlotName)
+        {
+            CloudStorageAccount account = CloudStorageAccount.Parse(storageString);
+            CloudBlobClient client = account.CreateCloudBlobClient();
+            CloudBlobContainer blobContainer = client.GetContainerReference("azure-webjobs-secrets");
+            await blobContainer.CreateIfNotExistsAsync();
+
+            string[] files = Directory.GetFiles(secretsPath);
+
+            if (files.Length > 0)
             {
-                return new FileSystemSecretsRepository(webHostSettings.SecretsPath);
+                foreach (string file in files)
```

Comment:
To be more efficient, don't do this serially. Kick off the tasks (at least for a batch of files at a time) and await on the batch result.